man/man1/htcrystalball.1
.Dd February 9, 2021
.Dt HTCRYSTALBALL 1
.Os \" Current operating system.
.
.Sh NAME
.Nm htcrystalball
.Nd calculate the runtime and maximum number of jobs that can run on an HTCondor pool
.
.Sh SYNOPSIS
.Nm
.Op Fl h
|
.Fl Fl cpu Ar num
.Fl Fl ram Ar size
.Op Fl g Ar num
.Op Fl d Ar size
.Op Fl j Ar num
.Op Fl t Ar time
.Op Fl m Ar num
.Op Fl f Ar path
.Op Fl v
.
.Sh DESCRIPTION
.Nm
calculates how many jobs
.Pq of a user\[hy]specified number and size
can run on an HTCondor pool.
It also can estimate runtime
.Pq core hours and wall time
and node\[hy]specific matching
.Pq aiding those with node\[hy]level licensing restrictions .
.Pp
The arguments are as follows:
.Bl -tag -width Ds
.
.It Fl c | Fl Fl cpu Ar number
The number of CPU cores per job.
.
.It Fl r | Fl Fl ram Ar size
The amount of RAM per job, including a unit
.Pq e.g. 10G .
.
.It OPTIONS
.
.It Fl h | Fl Fl help
Prints a summary of
.Nm Ap s
command\[hy]line options and exits.
.
.It Fl g | Fl Fl gpu Ar number
The number of GPUs per job.
.
.It Fl d | Fl Fl disk Ar size
The disk space per job, including a unit
.Pq e.g. 50G .
.
.It Fl j | Fl Fl jobs Ar number
The number of jobs to be executed.
.
.It Fl t | Fl Fl time Ar time
The estimated time for one job to be executed, including unit
.Pq e.g. 1h .
.
.It Fl m | Fl Fl maxnodes Ar number
The maximum number of nodes jobs can be executed on.
Sometimes necessary due to software license restrictions.
.
.It Fl f | Fl Fl file Ar path
The path to a condor submit-file for parsing resource requirements to use instead of typed ones for CPU, GPU, RAM and DISK.
.
.It Fl v | Fl Fl verbose
Prints a table listing each node, its resources, and proposed usage.
.El
.
.Sh Units
.Ss Storage
Valid storage units are
.Pq case insensitive :
.Bl -tag -width 4n -offset 4n
.It Ar K
kilobytes
.Pq or Ar KB , KiB
.It Ar M
megabytes
.Pq or Ar MB , MiB
.It Ar G
gigabytes
.Pq or Ar GB , GiB
.It Ar T
terabytes
.Pq or Ar TB , TiB
.It Ar P
petabytes
.Pq or Ar PB , PiB
.El
.
.Ss Time
Valid time units are
.Pq case insensitive :
.Bl -tag -width 4n -offset 4n
.It Ar d
days
.It Ar h
hours
.It Ar m
minutes
.It Ar s
seconds
.El
.
.Sh EXIT STATUS
.Ex -std
.
.Sh EXAMPLES
Basic output:
.Dl htcrystalball \-\-cpu 1 \-\-ram 7500M \-\-jobs 1
.
Verbose:
.Dl htcrystalball \-\-cpu 1 \-\-ram 7500M \-\-jobs 1 \-\-verbose
.
Big job with runtime projection:
.Dl htcb \-\-cpu 16 \-\-ram 16G \-\-disk 100G \-\-jobs 20 \-\-time 5h
.
Limit to a maximum number of nodes:
.Dl htcb \-\-cpu 16 \-\-ram 16G \-\-disk 100G \-\-maxnodes 2 \-\-jobs 20 \-\-time 5h
.
GPU job:
.Dl htcb \-\-cpu 1 \-\-gpu 1 \-\-ram 8G \-\-disk 64G \-\-jobs 10 \-\-time 2h
.
.Sh SEE ALSO
The
.Nm
repository is available at
.Lk https://github.com/psyinfra/HTCrystalBall .
Bug reports, patches, and constructive input are always welcome.
.
.Sh AUTHORS
.Nm
was created by
.An Jona Marcus Fischer
.Mt j.fischer@fz\[hy]juelich.de .
.
.Sh COPYRIGHT
.Nm
is released under the
.Qq ISC License .
See the LICENSE file for more information.


htcrystalball.py
#!/usr/bin/env python3

"""Gives users a preview on how and where they can execute their HTcondor compatible scripts."""
import argparse
import json
import re
import logging
import my_modules.check_slots as slot_checker
import fetch_slots as sloth

# External (root level) logging level
logging.basicConfig(level=logging.ERROR)

# Internal logging level
logger = logging.getLogger('crystal_balls')
logger.setLevel(level=logging.DEBUG)

SLOTS_CONFIGURATION = "config/slots.json"


def validate_storage_size(arg_value: str) -> str:
    """
    Defines and checks valid storage inputs.

    Args:
        arg_value: The given storage input string

    Returns:
        The valid storage string or raises an exception if it doesn't match the regex.
    """

    pat = re.compile(r"^[0-9]+([kKmMgGtTpP]i?[bB]?)$")

    if not pat.match(arg_value):
        logger.error(f'Invalid storage value given: {arg_value}')
        raise argparse.ArgumentTypeError
    return arg_value


def validate_duration(arg_value: str) -> str:
    """
    Defines and checks valid time inputs.

    Args:
        arg_value: The given duration input string

    Returns:
        The valid duration string or raises an exception if it doesn't match the regex.
    """
    pat = re.compile(r"^([0-9]+([dDhHmMsS]?))?$")

    if not pat.match(arg_value):
        logger.error(f'Invalid time value given: {arg_value}')
        raise argparse.ArgumentTypeError
    return arg_value


def split_number_unit(user_input: str) -> [float, str]:
    """
    Splits the user input for storage sizes into number and storage unit.
    If no value or unit is given, the unit is set to GiB.

    Args:
        user_input: The given number string

    Returns:
        The amount and unit string separated in a list.
    """
    if not user_input:
        return [0.0, "GiB"]

    splitted = re.split(r'(\d*\.?\d+)', user_input.replace(' ', ''))

    amount = float(splitted[1])
    if splitted[2] == "":
        unit = "GiB"
        logger.info("No storage unit given, using GiB as default.")
    else:
        unit = splitted[2]

    return [amount, unit]


def split_duration_unit(user_input: str) -> [float, str]:
    """
    Splits the user input for time into number and time unit.
    If no value or unit is given, the unit is set to minutes.

    Args:
        user_input: The given duration string

    Returns:
        The duration and unit string separated in a list.
    """
    if user_input == "" or user_input is None:
        return [0.0, "min"]

    splitted = re.split(r'(\d*\.?\d+)', user_input.replace(' ', ''))

    amount = float(splitted[1])
    if splitted[2] == "":
        unit = "min"
        logger.info("No duration unit given, using MIN as default.")
    else:
        unit = splitted[2]

    return [amount, unit]


def calc_to_bin(number: float, unit: str) -> float:
    """
    Converts a storage value to GiB and accounts for base2 and base10 units.

    Args:
        number: The storage size number
        unit: The storage unit string

    Returns:
        The storage size number converted to GiB
    """
    unit_indicator = unit.lower()
    if unit_indicator in ("kb", "k", "kib"):
        return number / (10 ** 6)
    if unit_indicator in ("mb", "m", "mib"):
        return number / (10 ** 3)
    if unit_indicator in ("tb", "t", "tib"):
        return number * (10 ** 3)
    if unit_indicator in ("pb", "p", "pib"):
        return number * (10 ** 6)
    return number


def calc_to_min(number: float, unit: str) -> float:
    """
    Converts a time value to minutes, according to the given unit.

    Args:
        number: The duration number
        unit: The duration unit string

    Returns:
        The duration number converted to minutes
    """
    unit_indicator = unit.lower()
    if unit_indicator in ("d", "dd"):
        return number * 24 * 60
    if unit_indicator in ("h", "hh"):
        return number * 60
    if unit_indicator in ("s", "ss"):
        return number / 60
    return number


def define_environment():
    """
    Defines the command line arguments and required formats.

    Returns:

    """
    parser = argparse.ArgumentParser(
        description="To get a preview for any job you are trying to execute using "
                    "HTCondor, please pass at least the number of CPUs and "
                    "the amount of RAM "
                    "(including units eg. 100MB, 90M, 10GB, 15G) to this script "
                    "according to the usage example shown above. For JOB Duration please "
                    "use d, h, m or s", prog='htcrystalball.py',
        usage='%(prog)s -c CPU -r RAM [-g GPU] [-d DISK] [-j JOBS] [-d DURATION] [-v]',
        epilog="PLEASE NOTE: HTCondor always uses binary storage "
               "sizes (1 GiB = 1024 MiB, 1 GB = 1000 MB), so inputs "
               "will automatically be treated that way.")
    parser.add_argument("-v", "--verbose", help="Print extended log to stdout",
                        action='store_true')
    parser.add_argument("-c", "--cpu", help="Set number of requested CPU Cores",
                        type=int, required=True)
    parser.add_argument("-g", "--gpu", help="Set number of requested GPU Units",
                        type=int)
    parser.add_argument("-j", "--jobs", help="Set number of jobs to be executed",
                        type=int)
    parser.add_argument("-t", "--time", help="Set the duration for one job "
                                             "to be executed", type=validate_duration)
    parser.add_argument("-d", "--disk", help="Set amount of requested disk "
                                             "storage", type=validate_storage_size)
    parser.add_argument("-r", "--ram", help="Set amount of requested memory "
                                            "storage", type=validate_storage_size, required=True)
    parser.add_argument("-m", "--maxnodes", help="Set maximum of nodes to "
                                                 "run jobs on", type=int)

    cmd_parser = parser.parse_args()
    return cmd_parser


def define_slots() -> dict:
    """
    Loads the slot configuration.

    Returns:

    """
    with open(SLOTS_CONFIGURATION) as config_file:
        data = json.load(config_file)
    return data["slots"]


def filter_slots(slots: dict, slot_type: str) -> list:
    """
    Filters the slots stored in a dictionary according to their type.

    Args:
        slots: Dictionary of slots
        slot_type: requested Slot Type for filtering

    Returns:
        A filtered dictionary of slots
    """
    res = []
    for node in slots:
        for slot in node["slot_size"]:
            if slot["SlotType"] == slot_type:
                slot["UtsnameNodename"] = node["UtsnameNodename"]
                res.append(slot)
    return res


def prepare_checking(cpu: int, gpu: int, ram: str, disk: str,
                     jobs: int, job_duration: str, maxnodes: int, verbose: bool) -> bool:
    """
    Loads the Slot configuration, handles storage and time inputs,
    and invokes the checking for given job request if the request is valid.

    Args:
        cpu: User input of CPU cores
        gpu: User input of GPU units
        ram: User input of the amount of RAM
        disk: User input of the amount of disk space
        jobs: User input of the number of similar jobs
        job_duration: User input of the duration time for a single job
        maxnodes:
        verbose:

    Returns:
        If all needed parameters were given
    """

    slot_config = define_slots()
    static_slts = filter_slots(slot_config, "static")
    dynamic_slts = filter_slots(slot_config, "dynamic")
    gpu_slts = filter_slots(slot_config, "gpu")

    [ram, ram_unit] = split_number_unit(ram)
    ram = calc_to_bin(ram, ram_unit)
    [disk, disk_unit] = split_number_unit(disk)
    disk = calc_to_bin(disk, disk_unit)

    [job_duration, duration_unit] = split_duration_unit(job_duration)
    job_duration = calc_to_min(job_duration, duration_unit)

    if cpu == 0:
        logger.warning("No number of CPU workers given --- ABORTING")
    elif ram == 0.0:
        logger.warning("No RAM amount given --- ABORTING")
    else:
        slot_checker.check_slots(static_slts, dynamic_slts, gpu_slts, cpu, ram, disk, gpu,
                                 jobs, job_duration, maxnodes, verbose)
        return True

    return False


if __name__ == "__main__":
    CMD_ARGS = define_environment()
    cpu_workers = CMD_ARGS.cpu
    if cpu_workers is None:
        cpu_workers = 0
    gpu_workers = CMD_ARGS.gpu
    if gpu_workers is None:
        gpu_workers = 0

    ram_amount = CMD_ARGS.ram
    disk_space = CMD_ARGS.disk

    job_amount = CMD_ARGS.jobs
    if job_amount is None:
        job_amount = 1
    job_duration = CMD_ARGS.time

    matlab_nodes = CMD_ARGS.maxnodes
    if matlab_nodes is None:
        matlab_nodes = 0

    # fetch current slot configuration
    sloth.run()

    prepare_checking(cpu_workers, gpu_workers, ram_amount, disk_space,
                     job_amount, job_duration, matlab_nodes, CMD_ARGS.verbose)
